connectors/cache/fallback.go

```diff
@@ -0,0 +1,222 @@
+
+	// TODO serializing dosa.Condition array? conditions could be any order
+	keysMap := rangeQuery{
+		Conditions: columnConditions,
+		Token:      token,
+		Limit:      limit,
+	}
+	cacheKey, _ := c.encoder.Encode(keysMap)
+	adaptedEi := adaptToKeyValue(ei)
+
+	if sourceErr == nil {
+		w := func() error {
+			rangeResults := rangeResults{
+				TokenNext: sourceToken,
+				Rows:      sourceRows,
+			}
+			cacheValue, _ := c.encoder.Encode(rangeResults)
+			newValues := map[string]dosa.FieldValue{
+				key:   cacheKey,
+				value: cacheValue,
+			}
+			return c.fallback.Upsert(ctx, adaptedEi, newValues)
+		}
+		_ = c.cacheWrite(w)
+
+		return sourceRows, sourceToken, sourceErr
+	}
+	value, err := c.getValueFromFallback(ctx, adaptedEi, cacheKey)
+	if err != nil {
+		return sourceRows, sourceToken, sourceErr
+	}
+	unpack := rangeResults{}
+	err = c.encoder.Decode(value, &unpack)
+	if err != nil {
+		return sourceRows, sourceToken, sourceErr
+	}
+	return unpack.Rows, unpack.TokenNext, err
+}
+
+// Scan returns scan result from origin.
+func (c *Connector) Scan(ctx context.Context, ei *dosa.EntityInfo, minimumFields []string, token string, limit int) ([]map[string]dosa.FieldValue, string, error) {
+	// Scan will just call range with no conditions
+	return c.Range(ctx, ei, nil, minimumFields, token, limit)
+}
+
+// Remove deletes an entry
+func (c *Connector) Remove(ctx context.Context, ei *dosa.EntityInfo, keys map[string]dosa.FieldValue) error {
+	w := func() error {
+		cacheKey := createCacheKey(ei, keys, c.encoder)
+		adaptedEi := adaptToKeyValue(ei)
+		return c.fallback.Remove(ctx, adaptedEi, map[string]dosa.FieldValue{key: cacheKey})
+	}
+	_ = c.cacheWrite(w)
+
+	return c.origin.Remove(ctx, ei, keys)
+}
+
+func (c *Connector) getValueFromFallback(ctx context.Context, ei *dosa.EntityInfo, keyValue []byte) ([]byte, error) {
+	response, err := c.fallback.Read(ctx, ei, map[string]dosa.FieldValue{key: keyValue}, dosa.All())
+	if err != nil {
+		return nil, err
+	}
+
+	// unpack the value
+	cacheValue, ok := response[value].([]byte)
+	if !ok {
+		return nil, errors.New("No value in cache for key")
+	}
+	return cacheValue, nil
+}
+
+func (c *Connector) setSynchronousMode(sync bool) {
+	c.synchronous = sync
+}
+
+func (c *Connector) cacheWrite(w func() error) error {
+	if c.synchronous {
+		return w()
+	}
+	go func() { _ = w() }()
+	return nil
+}
+
+func adaptToKeyValue(ei *dosa.EntityInfo) *dosa.EntityInfo {
+	adaptedEi := &dosa.EntityInfo{}
+	adaptedEi.Ref = ei.Ref
+	adaptedEi.Def = &dosa.EntityDefinition{
+		Name: ei.Def.Name,
+		Key: &dosa.PrimaryKey{
+			PartitionKeys: []string{key},
+		},
+		Columns: []*dosa.ColumnDefinition{
+			{Name: value, Type: dosa.Blob},
+			{Name: key, Type: dosa.Blob},
+		},
+	}
+	return adaptedEi
+}
+
+// used for single entry reads/writes
+func createCacheKey(ei *dosa.EntityInfo, values map[string]dosa.FieldValue, e Encoder) []byte {
```

Comment:
@eculver my plan is to order the keys alphabetically so that i can just use gob encoding instead of thrift. Thrift is pretty complicated as I would have to adapt everything over into the thrift objects manually